src/devices/cpu/sh/sh7021.cpp

```diff
@@ -1,164 +1,2169 @@
+}
+
+uint32_t sh7021_device::read_long(offs_t offset)
+{
+	const uint32_t area = (offset >> 24) & 7;
+	if (area == 6)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 11) & 3) + 1; // Consume cycles specified by A6LW
+	else if (area == 0 || area == 2)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 13) & 3) + 1; // Consume cycles specified by A02LW
+
+	return m_program->read_dword(offset & m_am);
+}
+
+uint16_t sh7021_device::decrypted_read_word(offs_t offset)
+{
+	const uint32_t area = (offset >> 24) & 7;
+	if (area == 6)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 11) & 3) + 1; // Consume cycles specified by A6LW
+	else if (area == 0 || area == 2)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 13) & 3) + 1; // Consume cycles specified by A02LW
+
+	return m_decrypted_program->read_word(offset);
+}
+
+void sh7021_device::write_byte(offs_t offset, uint8_t data)
+{
+	const uint32_t area = (offset >> 24) & 7;
+	if (area == 6)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 11) & 3) + 1; // Consume cycles specified by A6LW
+	else if (area == 0 || area == 2)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 13) & 3) + 1; // Consume cycles specified by A02LW
+
+	m_program->write_byte(offset & m_am, data);
+}
+
+void sh7021_device::write_word(offs_t offset, uint16_t data)
+{
+	const uint32_t area = (offset >> 24) & 7;
+	if (area == 6)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 11) & 3) + 1; // Consume cycles specified by A6LW
+	else if (area == 0 || area == 2)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 13) & 3) + 1; // Consume cycles specified by A02LW
+
+    m_program->write_word(offset & m_am, data);
+}
+
+void sh7021_device::write_long(offs_t offset, uint32_t data)
+{
+	const uint32_t area = (offset >> 24) & 7;
+	if (area == 6)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 11) & 3) + 1; // Consume cycles specified by A6LW
+	else if (area == 0 || area == 2)
+		m_sh2_state->icount -= ((m_bsc.wcr3 >> 13) & 3) + 1; // Consume cycles specified by A02LW
+
+	m_program->write_dword(offset & m_am, data);
+}
+
+void sh7021_device::recalc_irq()
+{
+	int irq = 0;
+	int vector = -1;
+
+	// Serial IRQs
+	for (uint32_t i = 0; i < 2; ++i)
 	{
-		logerror("Warning: SH7021 DMA enables non auto-request transfer\n");
-		return;
+		int sci_level = (i == 1 ? ((m_ipre >> 12) & 0xf) : (m_iprd & 0xf));
+		if (sci_level == 0)
+		{
+			continue;
+		}
+
+		if (BIT(m_sci[i].scr, 7) && BIT(m_sci[i].ssr, 7))
+		{
+			// TxI
+			irq = sci_level;
+			vector = 102 + i * 4;
+			LOGMASKED(LOG_DMA_WR, "SCI Tx interrupt %d is ready, level %d vector %d\n", i, irq, vector);
+		}
+		else if (BIT(m_sci[i].scr, 2) && BIT(m_sci[i].ssr, 2))
+		{
+			// TEI
+			irq = sci_level;
+			vector = 103 + i * 4;
+			LOGMASKED(LOG_DMA_WR, "SCI TE interrupt %d is ready, level %d vector %d\n", i, irq, vector);
+		}
 	}
 
-	// channel enable & master enable
-	if((m_dma[ch].chcr & 1) == 0 || (m_dmaor & 1) == 0)
+	// Timer IRQs
+	for (uint32_t i = 0; i < 5; ++i)
+	{
+		int level;
+		if ((m_itu.timer[i].tier & m_itu.timer[i].tsr) & 7)
+		{
```

Comment:
Does `level` need to be declared outside the `if` block?  It seems to only be used inside.